File: TreeKalk/minheap.cpp

#include "minheap.h"


minHeap::minHeap():heap(){setTipo(QString("minHeap"));}
minHeap::minHeap(const minHeap& b,QString q):heap(b.getRadice()->copy(),q){setTipo(QString("minHeap"));}
minHeap::minHeap(nodo* n,const QString& q):heap(n,q){setTipo(QString("minHeap"));}

minHeap::~minHeap()
{
    //cout << "~alberobinario()\n";
}

bool minHeap::operator ==(const minHeap& b)const
{
    return heap::operator ==(b);
}
bool minHeap::operator !=(const minHeap& b)const
{
    return !(operator ==(b));
}
minHeap& minHeap::operator =(const minHeap& b)
{
    if(this!=&b)
    {
        delete getRadice();
        setRoot(b.getRadice()->copy());
    }
    return *this;
}
void minHeap::heapify(binary* b)
{
    if(b)
    {
        binary* min=b;
        if(b->left() && b->left()->getInfo()<b->getInfo())
            min=b->left();
        if(b->right() && b->right()->getInfo()<min->getInfo())
            min=b->right();
        if(min!=b)
        {
            double t=b->getInfo();
            b->setInfo(min->getInfo());
            min->setInfo(t);
            heapify(min);
        }
    }
}
void minHeap::heapifyUp(binary* b)
{
    if(b && b->getPadre() && b->getInfo()<b->getPadre()->getInfo())
    {
        double t=b->getInfo();
        b->setInfo(b->getPadre()->getInfo());
        b->getPadre()->setInfo(t);
        heapifyUp(b->getPadre());
    }
}

void minHeap::insert(double i)
{
    if(!getRadice())
        setRoot(new binary(i,NULL,NULL,NULL));
    else
    {
        int dim=dimensione(getRadice());
        QString b=converti(dim+1);
        binary* temp=getRadice();
        int j=1;

        while(j<b.size()-1)
        {
            temp=( b.mid(j,1)=="0" )?temp->left():temp->right();
            j++;
        }
        if(b.mid(b.size()-1,1)=="0")
        {
            temp->setLeft(new binary(i,temp,NULL,NULL));
            heapifyUp(temp->left());
        }
        else if(b.mid(b.size()-1,1)=="1")
        {
            temp->setRight(new binary(i,temp,NULL,NULL));
            heapifyUp(temp->right());
        }
    }

}
void minHeap::cancella(nodo* n)               // cancella il nodo specifico dall'albero
{
    if(n && !contenuto(n))
        throw std::string("Nodo non contenuto.");
    binary * temp=dynamic_cast<binary*>(ultimo());
    if(temp==getRadice())
    {
        delete getRadice();
        setRoot(NULL);
    }
    else
    {
        if(temp && temp->getPadre())
        {
            if(temp==temp->getPadre()->right())
                temp->getPadre()->setRight(NULL);
            else
                temp->getPadre()->setLeft(NULL);
            temp->setPadre(NULL);
        }
        double i=n->getInfo();
        n->setInfo(temp->getInfo());
        if(n->getInfo()<i)
            heapifyUp(dynamic_cast<binary*>(n));
        else
            heapify(dynamic_cast<binary*>(n));
        if(temp)
            delete temp;
    }
}
void minHeap::cancella(double i)              // cancella la prima ricorrenza di i in ordine prefisso
{
    if(!contenuto(i))
        throw std::string("Info non contenuta.");
    cancella(trovaPrimaRicorrenza(i));
}
bool minHeap::search(double i)const           // ritorna true se trova i nell'albero
{
    return contenuto(i);
}
minHeap* minHeap::copy()                       // ritorna una copia dell'albero
{
    return new minHeap(getRadice()->copy(),getNome());
}
void minHeap::modifica(nodo* n,double info)   // modifica l'info di un nodo dell'albero
{
    if(info<n->getInfo())
    {
        n->setInfo(info);
        heapifyUp(dynamic_cast<binary*>(n));
    }
    else if(info<n->getInfo())
    {
        n->setInfo(info);
        heapify(dynamic_cast<binary*>(n));
    }
}
void minHeap::massimoHelp(binary* n,double& max)const                 // ritorna il minimo dell'albero
{
    if(n->getInfo()>max)
        max=n->getInfo();
    if(n->left())
        massimoHelp(n->left(),max);
    if(n->right())
        massimoHelp(n->right(),max);
}
double minHeap::minimo()const                 // ritorna il minimo dell'albero
{

    if(!getRadice())
        throw std::string("Albero vuoto.");
    return getRadice()->getInfo();
}
double minHeap::massimo()const
{
    if(!getRadice())
        throw std::string("Albero vuoto.");
    double max=getRadice()->getInfo();
    massimoHelp(getRadice(),max);
    return max;
}
QString minHeap::ordina()const
{
    binary* temp=getRadice()->copy();
    minHeap* h=new minHeap();
    h->setRoot(temp);
    QString q="[";
    while(h->dimensione(h->getRadice())>0)
    {
        q=q+QString::number(h->getRadice()->getInfo()) + ",";
        h->cancella(h->getRadice());
    }
    q.remove(q.size()-1,1);
    q+="]";
    return q;
}


binary* minHeap::operazioneHelp(nodo* n1,nodo* n2,char tipo)const
{
    minHeap* t=new minHeap();
    vector<double> dati,dati1;
    getDati(n1,dati);
    getDati(n2,dati1);
    vector<double>::const_iterator it=dati.begin(),it1=dati1.begin();
    for(int i=0;i<dati.size();i++)
    {
        if(tipo=='+')t->insert((*it)+(*it1));
        else if(tipo=='-')
            t->insert((*it)-(*it1));
        else if(tipo=='*')
            t->insert((*it)*(*it1));
        else if(tipo=='/')
        {
            if(*it1==0)
            {
                delete t;
                throw QString("Operazione non valida.Divisione per 0.");
            }
            t->insert((*it)/(*it1));
        }
        else
            throw QString("Operazione non valida");
        it++;it1++;
    }
    return t->getRadice();
}

minHeap* minHeap::operazione(Albero* a,char tipo)const
{
    if(!alberiUguali(a))
        throw QString("Alberi di dimensione diversa.Impossibile eseguire l'operazione.");
    minHeap* temp=new minHeap();
    temp->setRoot(operazioneHelp(getRadice(),a->getRadice(),tipo));
    return temp;
}
minHeap* minHeap::unione(Albero* a)const
{
    vector<double> dati;
    getDati(getRadice(),dati);
    getDati(a->getRadice(),dati);
    minHeap* ris=new minHeap();
    for(vector<double>::const_iterator it=dati.begin();it!=dati.end();it++)
        ris->insert(*it);
    return ris;
}
minHeap* minHeap::intersezione(Albero* a)const
{
    vector<double> dati1,dati2,dati3;
    getDati(getRadice(),dati1);
    getDati(a->getRadice(),dati2);
    for(vector<double>::const_iterator it=dati1.begin();it!=dati1.end();it++)
        if(find(dati2.begin(), dati2.end(), *it) != dati2.end())
        {
            vector<double>::iterator it1=dati2.begin();
            while(*it1!=*it)
                it1++;
            dati2.erase(it1);
            dati3.push_back(*it);
        }

    minHeap* ris=new minHeap();
    for(vector<double>::const_iterator it=dati3.begin();it!=dati3.end();it++)
        ris->insert(*it);
    return ris;
}
minHeap* minHeap::operator+(Albero* a)const
{
    return operazione(a,'+');
}
minHeap* minHeap::operator-(Albero* a)const
{
    return operazione(a,'-');
}
minHeap* minHeap::operator*(Albero* a)const
{
    return operazione(a,'*');
}
minHeap* minHeap::operator/(Albero* a)const
{
    return operazione(a,'/');
}

File: java/minHeap.java

import java.util.*;


public class minHeap extends heap
{

    public minHeap()
    {
        super();
    }
    public minHeap(minHeap m)
    {
        super(m.getRadice().copy());
    }
    public minHeap(nodo n)
    {
        super(n);
    }
    public void heapify(binary b)
    {
        if(b!=null)
        {
            binary min=b;
            if(b.left()!=null && b.left().getInfo()<b.getInfo())
                min=b.left();
            if(b.right()!=null && b.right().getInfo()<min.getInfo())
                min=b.right();
            if(min!=b)
            {
                double t=b.getInfo();
                b.setInfo(min.getInfo());
                min.setInfo(t);
                heapify(min);
            }
        }
    }
    public void heapifyUp(binary b)
    {
        if(b!=null && b.getPadre()!=null && b.getInfo()<b.getPadre().getInfo())
        {
            double t=b.getInfo();
            b.setInfo(b.getPadre().getInfo());
            b.getPadre().setInfo(t);
            heapifyUp(b.getPadre());
        }
    }
    public void insert(double i)
    {
        if(getRadice()==null)
            setRoot(new binary(i,null,null,null));
        else
        {
            int dim=dimensione(getRadice());
            int j=1;
            String b=converti(dim+1);
            binary temp=getRadice();
            while(j<b.length()-1)
            {
                if(b.charAt(j)=='0')
                    temp=temp.left();
                else if(b.charAt(j)=='1')
                    temp=temp.right();
                j++;
            }
            if(b.charAt(b.length()-1)=='0')
            {
                temp.setLeft(new binary(i,temp,null,null));
                heapifyUp(temp.left());
            }
            else if(b.charAt(b.length()-1)=='1')
            {
                temp.setRight(new binary(i,temp,null,null));
                heapifyUp(temp.right());
            }
        }
    }
    public void cancella(nodo n)
    {
        binary temp=(binary)ultimo();
        if(temp==getRadice())
            setRoot(null);
        else
        {
            if(temp!=null && temp.getPadre()!=null)
            {
                if(temp==temp.getPadre().right())
                    temp.getPadre().setRight(null);
                else
                    temp.getPadre().setLeft(null);
                temp.setPadre(null);
            }
            double i=n.getInfo();
            n.setInfo(temp.getInfo());
            if(n.getInfo()<i)
                heapifyUp((binary)n);
            else
                heapify((binary)n);
        }
    }
    public void cancella(double i)
    {
        cancella(trovaPrimaRicorrenza(i));
    }
    public boolean search(double i)
    {
        return contenuto(i);
    }
    public minHeap copy()
    {
        minHeap m=new minHeap();
        m.setRoot((binary)getRadice().copy());
        return m;
    }
    public void modifica(nodo n,double info)
    {
        n.setInfo(info);
        if(info<n.getInfo())
            heapifyUp((binary)n);
        else if(info>n.getInfo())
            heapify((binary)n);
    }
    public double massimoHelp(binary n)
    {
        double max=0;
        if(n.getInfo()>max)
            max=n.getInfo();
        if(n.left()!=null)
            if(massimoHelp(n.left())>max)
                max=massimoHelp(n.left());
        if(n.right()!=null)
            if(massimoHelp(n.right())>max)
                max=massimoHelp(n.right());
        return max;
    }
    public double minimo()
    {
        return getRadice().getInfo();
    }
    public double massimo()
    {
        return massimoHelp(getRadice());
    }
    public String ordina()//////////////////////////////////////////////////
    {
        binary temp=getRadice().copy();
        minHeap h=new minHeap();
        h.setRoot(temp);
        String q="";
        String r="";
        while(h.dimensione(h.getRadice())>1)
        {
            q=q+","+temp.getInfo();
            h.cancella(temp.getInfo());
        }
        q=q+","+temp.getInfo();
        q=q.substring(1,q.length());
        q="["+q+"]";
        return q;
    }
    public binary operazioneHelp(nodo n1,nodo n2,char tipo)throws exc
    {
	minHeap t=new minHeap();
	List dati=new ArrayList();
	List dati1=new ArrayList();
	getDati(n1,dati);
	getDati(n2,dati1);
	if(tipo=='/' && dati1.contains(0))
            throw new exc("Operazione non valida divisione per 0");
	for(int i=0 ; i<dati.size();i++)
	{
            if(tipo=='+')
                t.insert((double)dati.get(i)+(double)dati1.get(i));
            else if(tipo=='-')
                t.insert((double)dati.get(i)-(double)dati1.get(i));
            else if(tipo=='*')
                t.insert((double)dati.get(i)*(double)dati1.get(i));
            else if(tipo=='/')
                t.insert((double)dati.get(i)/(double)dati1.get(i));
            else
                throw new exc("Operazione non valida");
        }
        return t.getRadice();
    }
    public minHeap operazione(Albero a,char tipo)
    {
        minHeap temp=new minHeap();
        if(alberiUguali(a)==true)
            temp.setRoot(operazioneHelp(getRadice(),a.getRadice(),tipo));
            
        return temp;
    }
    public minHeap unione(Albero a)
    {
        List dati=new ArrayList();
        getDati(getRadice(),dati);
        getDati(a.getRadice(),dati);
        minHeap ris=new minHeap();
        for(int i=0;i<dati.size();i++)
            ris.insert((double)dati.get(i));
        return ris;
    }
    public minHeap intersezione(Albero a)
    {
        List dati1=new ArrayList();
        List dati2=new ArrayList();
        List dati3=new ArrayList();
        getDati(getRadice(),dati1);
        getDati((binary)a.getRadice(),dati2);
        for(int it=0 ;it<dati1.size();it++)
            if(dati2.contains(dati1.get(it)))
            {
                dati3.add(dati1.get(it));
                dati2.remove(dati1.get(it));
            }

        minHeap ris=new minHeap();
        for(int it=0 ;it<dati3.size();it++)
            ris.insert((double)dati3.get(it));
        return ris;
    }
    public minHeap operatorpiu(Albero a)
    {
        return operazione(a,'+');
    }
    public minHeap operatormeno(Albero a)
    {
        return operazione(a,'-');
    }
    public minHeap operatorper(Albero a)
    {
        return operazione(a,'*');
    }
    public minHeap operatordiviso(Albero a)
    {
        return operazione(a,'/');
    }



    public static void main(String []args)
    {
        minHeap a=new minHeap();
        a.insert(3.0);
        a.insert(5);
        a.insert(1);
        a.insert(5);
        a.insert(8);
        System.out.println("\na :");
        a.getRadice().stampa();
        minHeap b=new minHeap();
        b.insert(-12);
        b.insert(1);
        b.insert(-30);
        b.insert(23);
        b.insert(200);
        System.out.println("\nb :");
        b.getRadice().stampa();
        System.out.println("\nmassimo di a :" + a.massimo());
        System.out.println("\nminimo di a :" + a.minimo());
        System.out.println("\na in ordine :" + a.ordina());
        System.out.println("\ncerco 5 in a :" + a.search(5));
        System.out.println("\ncerco 15 in a :" + a.search(15));
        System.out.println("\na+b :");
        a.operatorpiu(b).getRadice().stampa();
        System.out.println("\na-b :");
        a.operatormeno(b).getRadice().stampa();
        System.out.println("\na*b :");
        a.operatorper(b).getRadice().stampa();
        System.out.println("\na/b :");
        a.operatordiviso(b).getRadice().stampa();
        System.out.println("\na unito b :");
        a.unione(b).getRadice().stampa();
        System.out.println("\na intersecato b :");
        a.intersezione(b).getRadice().stampa();
        System.out.println("\ncancello 5 da a: " );
        a.cancella(5);
        a.getRadice().stampa();
        
    }








}
